File: app/src/test/kotlin/com/silent/vpn/policy/OtaGithubDiscoveryTest.kt
package com.silent.vpn.policy

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class OtaGithubDiscoveryTest {

    private val json = """
        {
          "api_base": "https://89-125-188-100.nip.io",
          "android": {
            "version": "1.0.166",
            "filename": "SilentVPN-release-1.0.166.apk",
            "size": 27778447,
            "download_url": "https://github.com/silentvpn3/silentvpn3.github.io/releases/download/v1.0.166/SilentVPN-release-1.0.166.apk"
          },
          "pc": {
            "version": "1.0.166",
            "filename": "Silent.VPN.Setup.1.0.166.exe",
            "size": 83072897,
            "download_url": "https://github.com/silentvpn3/silentvpn3.github.io/releases/download/v1.0.166/Silent.VPN.Setup.1.0.166.exe"
          },
          "linux": {
            "version": "1.0.166",
            "filename": "Silent.VPN.Setup.1.0.166.deb",
            "size": 118399176,
            "download_url": "https://github.com/silentvpn3/silentvpn3.github.io/releases/download/v1.0.166/Silent.VPN.Setup.1.0.166.deb"
          }
        }
    """.trimIndent()

    @Test
    fun `releases json url is github pages not hive ip`() {
        assertEquals(
            "https://silentvpn3.github.io/releases.json",
            OtaGithubDiscovery.RELEASES_JSON_URL,
        )
        assertTrue(!OtaGithubDiscovery.RELEASES_JSON_URL.contains("nip.io"))
        assertTrue(!OtaGithubDiscovery.RELEASES_JSON_URL.contains("89.125"))
    }

    @Test
    fun `165 android sees 166 from github pages`() {
        val got = OtaGithubDiscovery.parse(json, platform = "android", currentVersion = "1.0.165")
        val offer = got as OtaGithubDiscovery.Result.Available
        assertEquals("1.0.166", offer.version)
        assertEquals("SilentVPN-release-1.0.166.apk", offer.filename)
        assertEquals(27778447L, offer.size)
        assertTrue(offer.downloadUrl.contains("github.com/silentvpn3"))
    }

    @Test
    fun `same version is current so hive ip outage cannot hide up-to-date`() {
        val got = OtaGithubDiscovery.parse(json, platform = "android", currentVersion = "1.0.166")
        assertEquals(OtaGithubDiscovery.Result.Current("1.0.166"), got)
    }

    @Test
    fun `garbage or missing json falls through to api`() {
        assertEquals(OtaGithubDiscovery.Result.Unreadable, OtaGithubDiscovery.parse(null, "android", "1.0.165"))
        assertEquals(OtaGithubDiscovery.Result.Unreadable, OtaGithubDiscovery.parse("{", "android", "1.0.165"))
        assertEquals(OtaGithubDiscovery.Result.Unreadable, OtaGithubDiscovery.parse("{}", "mac", "1.0.165"))
    }

    @Test
    fun `pc linux and tv map onto landing keys`() {
        val pc = OtaGithubDiscovery.parse(json, "pc", "1.0.165") as OtaGithubDiscovery.Result.Available
        val linux = OtaGithubDiscovery.parse(json, "linux", "1.0.165") as OtaGithubDiscovery.Result.Available
        val tv = OtaGithubDiscovery.parse(json, "android_tv", "1.0.165") as OtaGithubDiscovery.Result.Available
        assertTrue(pc.downloadUrl.endsWith(".exe"))
        assertTrue(linux.downloadUrl.endsWith(".deb"))
        assertTrue(tv.downloadUrl.endsWith(".apk"))
        assertEquals("android", OtaGithubDiscovery.landingKey("android_tv"))
        assertEquals("mac", OtaGithubDiscovery.landingKey("darwin"))
        assertEquals("pc", OtaGithubDiscovery.landingKey("windows"))
    }
}

File: test/otaGithubDiscovery.test.js
const { describe, it } = require('node:test')
const assert = require('node:assert/strict')
const {
  RELEASES_JSON_URL,
  parseGithubOta,
  landingKey,
  otaDownloadCandidates,
} = require('../src/main/vpn/otaGithubDiscovery')

const JSON_BODY = JSON.stringify({
  api_base: 'https://89-125-188-100.nip.io',
  android: {
    version: '1.0.166',
    filename: 'SilentVPN-release-1.0.166.apk',
    size: 27778447,
    download_url: 'https://github.com/silentvpn3/silentvpn3.github.io/releases/download/v1.0.166/SilentVPN-release-1.0.166.apk',
  },
  pc: {
    version: '1.0.166',
    filename: 'Silent.VPN.Setup.1.0.166.exe',
    size: 83072897,
    download_url: 'https://github.com/silentvpn3/silentvpn3.github.io/releases/download/v1.0.166/Silent.VPN.Setup.1.0.166.exe',
  },
  linux: {
    version: '1.0.166',
    filename: 'Silent.VPN.Setup.1.0.166.deb',
    size: 118399176,
    download_url: 'https://github.com/silentvpn3/silentvpn3.github.io/releases/download/v1.0.166/Silent.VPN.Setup.1.0.166.deb',
  },
})

describe('otaGithubDiscovery', () => {
  it('uses GitHub Pages, not the hive nip.io', () => {
    assert.equal(RELEASES_JSON_URL, 'https://silentvpn3.github.io/releases.json')
    assert.equal(RELEASES_JSON_URL.includes('nip.io'), false)
  })

  it('offers 1.0.166 to a 1.0.165 PC without asking the hive', () => {
    const got = parseGithubOta(JSON_BODY, 'pc', '1.0.165')
    assert.equal(got.kind, 'available')
    assert.equal(got.version, '1.0.166')
    assert.match(got.download_url, /github\.com\/silentvpn3/)
  })

  it('treats same version as current', () => {
    const got = parseGithubOta(JSON_BODY, 'pc', '1.0.166')
    assert.equal(got.kind, 'current')
    assert.equal(got.version, '1.0.166')
  })

  it('is unreadable on garbage so the client does not invent a hive check', () => {
    assert.equal(parseGithubOta(null, 'pc', '1.0.165').kind, 'unreadable')
    assert.equal(parseGithubOta('{', 'pc', '1.0.165').kind, 'unreadable')
  })

  it('mac arches: arm64 and x64 each get their own DMG', () => {
    const body = JSON.stringify({
      mac: {
        version: '1.0.168',
        filename: 'Silent.VPN.Setup.1.0.168-x64.dmg',
        size: 1,
        download_url: 'https://github.com/silentvpn3/silentvpn3.github.io/releases/download/v1.0.168/Silent.VPN.Setup.1.0.168-x64.dmg',
        arches: {
          x64: {
            version: '1.0.168',
            filename: 'Silent.VPN.Setup.1.0.168-x64.dmg',
            size: 1,
            download_url: 'https://github.com/silentvpn3/silentvpn3.github.io/releases/download/v1.0.168/Silent.VPN.Setup.1.0.168-x64.dmg',
          },
          arm64: {
            version: '1.0.168',
            filename: 'Silent.VPN.Setup.1.0.168-arm64.dmg',
            size: 2,
            download_url: 'https://github.com/silentvpn3/silentvpn3.github.io/releases/download/v1.0.168/Silent.VPN.Setup.1.0.168-arm64.dmg',
          },
        },
      },
    })
    const arm = parseGithubOta(body, 'mac', '1.0.167', 'arm64')
    const intel = parseGithubOta(body, 'mac', '1.0.167', 'x64')
    assert.equal(arm.kind, 'available')
    assert.match(arm.download_url, /-arm64\.dmg/)
    assert.match(intel.download_url, /-x64\.dmg/)
  })

  it('maps windows/darwin/tv onto landing keys and never prefers hive download', () => {
    assert.equal(landingKey('windows'), 'pc')
    assert.equal(landingKey('darwin'), 'mac')
    assert.equal(landingKey('android_tv'), 'android')
    const urls = otaDownloadCandidates({
      githubUrl: 'https://github.com/silentvpn3/silentvpn3.github.io/releases/download/v1.0.166/app.exe',
      vpnUp: true,
      tunnelOrigin: 'http://10.66.66.1:8000',
      hiveDownloadPath: '/api/updates/download/pc',
    })
    assert.equal(urls[0].startsWith('https://github.com/'), true)
    assert.equal(urls.some((u) => u.includes('10.66.66.1')), false)
  })
})
